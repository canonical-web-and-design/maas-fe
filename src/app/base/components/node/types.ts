import type { CommonActionFormProps } from "app/base/types";
import type { Node } from "app/store/types/node";

export type NodeActionFormProps<E = null> = CommonActionFormProps<E> & {
  cleanup?: () => void;
  nodes: Node[];
  modelName: string;
};
