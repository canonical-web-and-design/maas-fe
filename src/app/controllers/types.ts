import type { ValueOf } from "@canonical/react-components";

import type { SidePanelContent, SetSidePanelContent } from "app/base/types";
import type { ControllerHeaderViews } from "app/controllers/constants";

export type ControllerSidePanelContent = SidePanelContent<
  ValueOf<typeof ControllerHeaderViews>
>;

export type ControllerSetSidePanelContent =
  SetSidePanelContent<ControllerSidePanelContent>;
