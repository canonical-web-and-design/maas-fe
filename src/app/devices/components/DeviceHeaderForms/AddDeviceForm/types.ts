import type {
  CreateParams,
  Device,
  DeviceIpAssignment,
} from "app/store/device/types";
import type { Domain } from "app/store/domain/types";
import type { Zone } from "app/store/zone/types";

type CreateParamsInterface = CreateParams["interfaces"][0];

export type AddDeviceInterface = {
  id: number;
  ip_address: NonNullable<CreateParamsInterface["ip_address"]>;
  ip_assignment: DeviceIpAssignment;
  mac: CreateParamsInterface["mac"];
  name: NonNullable<CreateParamsInterface["name"]>;
  subnet: NonNullable<CreateParamsInterface["subnet"]> | "";
};

export type AddDeviceValues = {
  domain: Domain["name"];
  hostname: Device["hostname"];
  interfaces: AddDeviceInterface[];
  zone: Zone["name"];
};
