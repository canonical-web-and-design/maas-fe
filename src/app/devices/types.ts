import type { ValueOf } from "@canonical/react-components";

import type { HeaderContent, SetHeaderContent } from "app/base/types";
import type { DeviceHeaderViews } from "app/devices/constants";

export type DeviceHeaderContent = HeaderContent<
  ValueOf<typeof DeviceHeaderViews>
>;

export type DeviceSetHeaderContent = SetHeaderContent<DeviceHeaderContent>;
