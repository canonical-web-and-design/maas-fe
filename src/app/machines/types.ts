import type { ValueOf } from "@canonical/react-components";

import type { MachineHeaderViews } from "./constants";

import type { HardwareType } from "app/base/enum";
import type {
  CommonActionFormProps,
  HeaderContent,
  SetHeaderContent,
} from "app/base/types";
import type {
  Machine,
  MachineEventErrors,
  SelectedMachines,
} from "app/store/machine/types";
import type { Script } from "app/store/script/types";

export type MachineHeaderContent = HeaderContent<
  ValueOf<typeof MachineHeaderViews>,
  {
    applyConfiguredNetworking?: Script["apply_configured_networking"];
    hardwareType?: HardwareType;
  }
>;

export type MachineSetHeaderContent = SetHeaderContent<MachineHeaderContent>;

export type MachineActionVariableProps = {
  machines?: Machine[];
  selectedMachines?: SelectedMachines | null;
  searchFilter?: string;
  selectedCount?: number | null;
  processingCount?: number;
  selectedCountLoading?: boolean;
};

export type MachineActionFormProps = Omit<
  CommonActionFormProps<MachineEventErrors>,
  "processingCount"
> &
  MachineActionVariableProps;
