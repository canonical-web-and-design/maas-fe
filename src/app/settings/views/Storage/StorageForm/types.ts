export type StorageFormValues = {
  default_storage_layout: string;
  disk_erase_with_quick_erase: boolean;
  disk_erase_with_secure_erase: boolean;
  enable_disk_erasing_on_release: boolean;
};
