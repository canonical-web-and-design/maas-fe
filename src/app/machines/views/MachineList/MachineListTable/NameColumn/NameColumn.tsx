import { memo } from "react";

import { Tooltip } from "@canonical/react-components";
import { useSelector } from "react-redux";
import { Link } from "react-router-dom-v5-compat";

import DoubleRow from "app/base/components/DoubleRow";
import NonBreakingSpace from "app/base/components/NonBreakingSpace";
import RowCheckbox from "app/base/components/RowCheckbox";
import machineSelectors from "app/store/machine/selectors";
import type { Machine, MachineMeta } from "app/store/machine/types";
import type { RootState } from "app/store/root/types";

type Props = {
  handleCheckbox?: () => void;
  selected: Machine[MachineMeta.PK][];
  showMAC?: boolean;
  systemId: Machine[MachineMeta.PK];
};

const generateFQDN = (machine: Machine, machineURL: string) => {
  return (
    <Link to={machineURL} title={machine.fqdn}>
      <strong data-testid="hostname">
        {machine.locked ? (
          <span title="This machine is locked. You have to unlock it to perform any actions.">
            <i className="p-icon--locked">Locked: </i>{" "}
          </span>
        ) : null}
        {machine.hostname}
      </strong>
      <small>.{machine.domain.name}</small>
    </Link>
  );
};

const generateIPAddresses = (machine: Machine) => {
  const ipAddresses: string[] = [];
  let bootIP;

  (machine.ip_addresses || []).forEach((address) => {
    let ip = address.ip;
    if (address.is_boot) {
      ip = `${ip} (PXE)`;
      bootIP = ip;
    }
    if (!ipAddresses.includes(ip)) {
      ipAddresses.push(ip);
    }
  });

  if (ipAddresses.length) {
    const ipAddressesLine = (
      <span
        data-testid="ip-addresses"
        title={ipAddresses.length === 1 ? ipAddresses[0] : ""}
      >
        {bootIP || ipAddresses[0]}
        {ipAddresses.length > 1 ? ` (+${ipAddresses.length - 1})` : null}
      </span>
    );

    if (ipAddresses.length === 1) {
      return ipAddressesLine;
    }
    return (
      <Tooltip
        position="btm-left"
        message={
          <>
            <strong>{ipAddresses.length} interfaces:</strong>
            <ul className="p-list u-no-margin--bottom">
              {ipAddresses.map((address) => (
                <li key={address}>{address}</li>
              ))}
            </ul>
          </>
        }
        positionElementClassName="p-double-row__tooltip-inner"
      >
        {ipAddressesLine}
      </Tooltip>
    );
  }
  return "";
};

const generateMAC = (machine: Machine, machineURL: string) => {
  return (
    <>
      <Link to={machineURL} title={machine.pxe_mac_vendor}>
        {machine.pxe_mac}
      </Link>
      {machine.extra_macs && machine.extra_macs.length > 0 ? (
        <Link to={machineURL}> (+{machine.extra_macs.length})</Link>
      ) : null}
    </>
  );
};

export const NameColumn = ({
  handleCheckbox,
  selected,
  showMAC,
  systemId,
}: Props): JSX.Element | null => {
  const machine = useSelector((state: RootState) =>
    machineSelectors.getById(state, systemId)
  );
  if (!machine) {
    return null;
  }
  const machineURL = `/${machine.link_type}/${machine.system_id}`;
  const primaryRow = showMAC
    ? generateMAC(machine, machineURL)
    : generateFQDN(machine, machineURL);
  const secondaryRow = !showMAC && generateIPAddresses(machine);

  return (
    <DoubleRow
      data-testid="name-column"
      primary={
        handleCheckbox ? (
          <RowCheckbox
            handleRowCheckbox={handleCheckbox}
            item={systemId}
            items={selected}
            inputLabel={primaryRow}
          />
        ) : (
          primaryRow
        )
      }
      // fallback to non-breaking space to keep equal height of all rows
      secondary={secondaryRow || <NonBreakingSpace />}
      secondaryClassName={
        handleCheckbox && "u-nudge--secondary-row u-align--left"
      }
    />
  );
};

export default memo(NameColumn);
