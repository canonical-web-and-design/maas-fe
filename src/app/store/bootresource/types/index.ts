export type {
  DeleteImageParams,
  FetchParams,
  OsystemParam,
  SaveOtherParams,
  SaveUbuntuCoreParams,
  SaveUbuntuParams,
} from "./actions";
export type {
  BaseImageFields,
  BootResource,
  BootResourceEventError,
  BootResourceFetchedArch,
  BootResourceFetchedImages,
  BootResourceFetchedRelease,
  BootResourceOtherImage,
  BootResourceState,
  BootResourceStatuses,
  BootResourceUbuntu,
  BootResourceUbuntuArch,
  BootResourceUbuntuCoreImage,
  BootResourceUbuntuRelease,
  BootResourceUbuntuSource,
} from "./base";
export {
  BootResourceAction,
  BootResourceMeta,
  BootResourceSourceType,
  BootResourceType,
} from "./enum";
export type {
  BootResourceFetchResponse,
  BootResourcePollResponse,
} from "./responses";
