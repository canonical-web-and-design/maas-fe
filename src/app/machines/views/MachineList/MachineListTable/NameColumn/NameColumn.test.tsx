import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { CompatRouter } from "react-router-dom-v5-compat";
import configureStore from "redux-mock-store";

import { NameColumn } from "./NameColumn";

import type { RootState } from "app/store/root/types";
import { NodeStatus } from "app/store/types/node";
import {
  modelRef as modelRefFactory,
  machine as machineFactory,
  machineState as machineStateFactory,
  rootState as rootStateFactory,
} from "testing/factories";

const mockStore = configureStore();

describe("NameColumn", () => {
  let state: RootState;
  beforeEach(() => {
    state = rootStateFactory({
      machine: machineStateFactory({
        loaded: true,
        items: [
          machineFactory({
            domain: modelRefFactory({
              name: "example",
            }),
            extra_macs: [],
            hostname: "koala",
            ip_addresses: [],
            pool: modelRefFactory(),
            pxe_mac: "00:11:22:33:44:55",
            status: NodeStatus.RELEASING,
            system_id: "abc123",
            zone: modelRefFactory(),
          }),
        ],
      }),
    });
  });

  it("can be locked", () => {
    state.machine.items[0].locked = true;
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find(".p-icon--locked").exists()).toBe(true);
  });

  it("can show the FQDN", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("a").text()).toEqual("koala.example");
  });

  it("can show a single ip address", () => {
    state.machine.items[0].ip_addresses = [{ ip: "127.0.0.1", is_boot: false }];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('[data-testid="ip-addresses"]').text()).toBe(
      "127.0.0.1"
    );
    // Doesn't show tooltip.
    expect(wrapper.find("Tooltip").exists()).toBe(false);
  });

  it("can show multiple ip addresses", () => {
    state.machine.items[0].ip_addresses = [
      { ip: "127.0.0.1", is_boot: false },
      { ip: "127.0.0.2", is_boot: false },
    ];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('[data-testid="ip-addresses"]').text()).toBe(
      "127.0.0.1"
    );
    expect(wrapper.find("Button").text()).toBe("+1");
    // Shows a tooltip.
    expect(wrapper.find("Tooltip").exists()).toBe(true);
  });

  it("can show a PXE ip address", () => {
    state.machine.items[0].ip_addresses = [{ is_boot: true, ip: "127.0.0.1" }];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('[data-testid="ip-addresses"]').text()).toBe(
      "127.0.0.1 (PXE)"
    );
  });

  it("doesn't show duplicate ip addresses", () => {
    state.machine.items[0].ip_addresses = [
      { ip: "127.0.0.1", is_boot: false },
      { ip: "127.0.0.1", is_boot: false },
    ];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find('[data-testid="ip-addresses"]').text()).toBe(
      "127.0.0.1"
    );
    // Doesn't show toolip.
    expect(wrapper.find("Tooltip").exists()).toBe(false);
  });

  it("can show a single mac address", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} showMAC={true} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("a").text()).toEqual("00:11:22:33:44:55");
  });

  it("can show multiple mac address", () => {
    state.machine.items[0].extra_macs = ["aa:bb:cc:dd:ee:ff"];
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} showMAC={true} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("a").length).toEqual(2);
    expect(wrapper.find("a").at(1).text()).toEqual(" (+1)");
  });

  it("can render a machine with minimal data", () => {
    state.machine.items[0] = machineFactory({
      domain: modelRefFactory({
        name: "example",
      }),
      hostname: "koala",
      system_id: "abc123",
    });
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("NameColumn").exists()).toBe(true);
  });

  it("can render a machine in the MAC state with minimal data", () => {
    state.machine.items[0] = machineFactory({
      domain: modelRefFactory({
        name: "example",
      }),
      hostname: "koala",
      system_id: "abc123",
    });
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} showMAC={true} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("NameColumn").exists()).toBe(true);
  });

  it("does not render checkbox if onToggleMenu not provided", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machines", key: "testKey" }]}
        >
          <CompatRouter>
            <NameColumn groupValue={null} systemId="abc123" />
          </CompatRouter>
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("Input").exists()).toBe(false);
  });
});
