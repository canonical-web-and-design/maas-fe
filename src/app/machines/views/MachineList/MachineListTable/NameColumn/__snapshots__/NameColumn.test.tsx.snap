// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameColumn renders 1`] = `
<NameColumn
  handleCheckbox={[MockFunction]}
  selected={Array []}
  systemId="abc123"
>
  <DoubleRow
    data-testid="name-column"
    primary={
      <RowCheckbox
        handleRowCheckbox={[MockFunction]}
        inputLabel={
          <Link
            title="test-machine-1.maas"
            to="/machine/abc123"
          >
            <strong
              data-testid="hostname"
            >
              koala
            </strong>
            <small>
              .
              example
            </small>
          </Link>
        }
        item="abc123"
        items={Array []}
      />
    }
    secondary={<NonBreakingSpace />}
    secondaryClassName="u-nudge--secondary-row u-align--left"
  >
    <div
      className="p-double-row"
    >
      <div
        className="p-double-row__rows-container"
      >
        <div
          className="p-double-row__primary-row"
        >
          <div
            className="p-double-row__primary-row-text u-truncate"
            data-testid="primary"
          >
            <RowCheckbox
              handleRowCheckbox={[MockFunction]}
              inputLabel={
                <Link
                  title="test-machine-1.maas"
                  to="/machine/abc123"
                >
                  <strong
                    data-testid="hostname"
                  >
                    koala
                  </strong>
                  <small>
                    .
                    example
                  </small>
                </Link>
              }
              item="abc123"
              items={Array []}
            >
              <Input
                checked={false}
                className="keep-label-opacity"
                id="mock-redux-js-nanoid-1"
                label={
                  <Link
                    title="test-machine-1.maas"
                    to="/machine/abc123"
                  >
                    <strong
                      data-testid="hostname"
                    >
                      koala
                    </strong>
                    <small>
                      .
                      example
                    </small>
                  </Link>
                }
                labelClassName="u-no-margin--bottom u-no-padding--top"
                onChange={[Function]}
                type="checkbox"
                wrapperClassName="u-no-margin--bottom u-nudge--checkbox"
              >
                <Field
                  className="u-no-margin--bottom u-nudge--checkbox"
                  forId="mock-redux-js-nanoid-1"
                  helpId="mock-nanoid-2"
                  isTickElement={true}
                  label=""
                  labelClassName="u-no-margin--bottom u-no-padding--top"
                  validationId="mock-nanoid-1"
                >
                  <div
                    className="p-form__group p-form-validation u-no-margin--bottom u-nudge--checkbox"
                  >
                    <div
                      className="p-form__control u-clearfix"
                    >
                      <CheckboxInput
                        aria-describedby={null}
                        aria-errormessage={null}
                        aria-invalid={false}
                        checked={false}
                        id="mock-redux-js-nanoid-1"
                        label={
                          <Link
                            title="test-machine-1.maas"
                            to="/machine/abc123"
                          >
                            <strong
                              data-testid="hostname"
                            >
                              koala
                            </strong>
                            <small>
                              .
                              example
                            </small>
                          </Link>
                        }
                        labelClassName="u-no-margin--bottom u-no-padding--top"
                        onChange={[Function]}
                      >
                        <CheckableInput
                          aria-describedby={null}
                          aria-errormessage={null}
                          aria-invalid={false}
                          checked={false}
                          id="mock-redux-js-nanoid-1"
                          indeterminate={false}
                          inputType="checkbox"
                          label={
                            <Link
                              title="test-machine-1.maas"
                              to="/machine/abc123"
                            >
                              <strong
                                data-testid="hostname"
                              >
                                koala
                              </strong>
                              <small>
                                .
                                example
                              </small>
                            </Link>
                          }
                          labelClassName="u-no-margin--bottom u-no-padding--top"
                          onChange={[Function]}
                        >
                          <label
                            className="u-no-margin--bottom u-no-padding--top p-checkbox"
                          >
                            <input
                              aria-describedby={null}
                              aria-errormessage={null}
                              aria-invalid={false}
                              aria-labelledby="mock-nanoid-3"
                              checked={false}
                              className="p-checkbox__input"
                              id="mock-redux-js-nanoid-1"
                              onChange={[Function]}
                              type="checkbox"
                            />
                            <span
                              className="p-checkbox__label"
                              id="mock-nanoid-3"
                            >
                              <Link
                                title="test-machine-1.maas"
                                to="/machine/abc123"
                              >
                                <a
                                  href="/machine/abc123"
                                  onClick={[Function]}
                                  title="test-machine-1.maas"
                                >
                                  <strong
                                    data-testid="hostname"
                                  >
                                    koala
                                  </strong>
                                  <small>
                                    .
                                    example
                                  </small>
                                </a>
                              </Link>
                            </span>
                          </label>
                        </CheckableInput>
                      </CheckboxInput>
                    </div>
                  </div>
                </Field>
              </Input>
            </RowCheckbox>
          </div>
        </div>
        <div
          className="p-double-row__secondary-row u-truncate u-nudge--secondary-row u-align--left"
          data-testid="secondary"
        >
          <NonBreakingSpace>
            Â 
          </NonBreakingSpace>
        </div>
      </div>
    </div>
  </DoubleRow>
</NameColumn>
`;
